releng_sop/koji_sign.py

```diff
@@ -0,0 +1,715 @@
+        return signed, unsigned
+
+    def write_signed_rpms_from_sigcache(self, rpm_info_list, sigkey, commit=False):
+        """
+        Reconstruct RPMs in koji from existing signed headers in sigcache.
+
+        :param rpm_info_list: List of koji rpm_info dictionaries
+        :type  rpm_info_list: list
+        :param sigkey: Sigkey
+        :type  sigkey: str
+        :param commit: Disable dry-run, apply changes for real.
+        :type  commit: bool=False
+        """
+        if commit:
+            self.koji_session.multicall = True
+
+        for rpm_info in rpm_info_list:
+            path = self._get_rpm_path(rpm_info, sigkey)
+
+            msg = "Writing RPM from '%s' sigcache: %s" % (sigkey, path)
+            self.log("info", msg, commit=commit)
+
+            if commit:
+                self.koji_session.writeSignedRPM(rpm_info, sigkey)
+
+        if commit:
+            self.koji_session.multiCall()
+
+    def _get_rpm_sighdr_sigkey(self, path):
+        """
+        Read header and sigkey from an RPM.
+
+        :param path: Path to a RPM package
+        :type  path: str
+        :return: (sighdr, sigkey)
+        :rtype:  tuple
+        """
+        # I/O is expensive, cache RPM headers and sigkeys
+        result = self._get_rpm_sighdr_sigkey_cache.get(path)
+        if result:
+            return result
+
+        sighdr = koji.rip_rpm_sighdr(path)
+        rawhdr = koji.RawHeader(sighdr)
+
+        sigpkt = rawhdr.get(koji.RPM_SIGTAG_GPG)
+        if not sigpkt:
+            sigpkt = rawhdr.get(koji.RPM_SIGTAG_PGP)
+
+        sigkey = ""
+        if sigpkt:
+            sigkey = koji.get_sigpacket_key_id(sigpkt)
+        sigkey = sigkey.lower()
+
+        result = (sighdr, sigkey)
+        self._get_rpm_sighdr_sigkey_cache[path] = result
+        return result
+
+    def split_rpm_info_list_by_size_and_files(self, rpm_info_list, max_size=500, max_files=20):
+        """
+        Split rpm_info_list into chunks by max file size or file count.
+
+        That should keep /tmp utilization in reasonable limits.
+
+        :param rpm_info_list: List of koji rpm_info dictionaries
+        :type  rpm_info_list: list
+        :param max_size: Maximal size of files to be signed at once
+        :type  max_size: int=500
+        :param max_files: Maximal count of files to be signed at once
+        :type  max_files: int=20
+        :return: [[rpm_info, ...], ...]
+        :rtype:  list
+        """
+        # make a copy, don't modify original
+        rpm_info_list = rpm_info_list[:]
+
+        while rpm_info_list:
+            size = 0
+            files = 0
+            result = []
+
+            while rpm_info_list:
+                rpm_info = rpm_info_list.pop(0)
+                result.append(rpm_info)
+                size += rpm_info["size"]
+                files += 1
+
+                if files and files >= max_files:
+                    break
+                if size and size >= max_size * 1024 ** 2:
```

Comment:
I'm thinking if it's necessary to check that ``files`` and ``size``} are non-empty, when the append into them is done just a line above = they cannot be empty (``rpm_info_list`` is guaranteed to be non-empty here, so ``pop()`` from it always return an element).